plugins/corsair/fu-corsair-device.c

```diff
@@ -0,0 +1,519 @@
+ */
+
+#include "config.h"
+
+#include <string.h>
+
+#include "fu-corsair-common.h"
+#include "fu-corsair-device.h"
+
+#define CORSAIR_VENDOR_INTERFACE_ID 1
+#define CORSAIR_CMD_SIZE	    64
+#define CORSAIR_TRANSACTION_TIMEOUT 2000
+#define CORSAIR_FIRST_CHUNK_SIZE    57
+#define CORSAIR_NEXT_CHUNKS_SIZE    61
+
+struct _FuCorsairDevice {
+	FuUsbDevice parent_instance;
+	FuCorsairDeviceKind device_kind;
+	guint8 epin;
+	guint8 epout;
+};
+G_DEFINE_TYPE(FuCorsairDevice, fu_corsair_device, FU_TYPE_USB_DEVICE)
+
+static gboolean
+fu_corsair_device_command(FuDevice *device, guint8 *data, guint timeout, GError **error)
+{
+	FuCorsairDevice *self = FU_CORSAIR_DEVICE(device);
+	GUsbDevice *usb_device = fu_usb_device_get_dev(FU_USB_DEVICE(device));
+	gsize actual_len = 0;
+	gboolean ret;
+
+	ret = g_usb_device_interrupt_transfer(usb_device,
+					      self->epout,
+					      data,
+					      CORSAIR_CMD_SIZE,
+					      &actual_len,
+					      timeout,
+					      NULL,
+					      error);
+	if (!ret) {
+		g_prefix_error(error, "failed to write command: ");
+		return FALSE;
+	}
+	if (actual_len != CORSAIR_CMD_SIZE) {
+		g_set_error(error,
+			    G_IO_ERROR,
+			    G_IO_ERROR_INVALID_DATA,
+			    "wrong size written: %" G_GSIZE_FORMAT,
+			    actual_len);
+		return FALSE;
+	}
+
+	memset(data, 0, CORSAIR_CMD_SIZE);
+
+	ret = g_usb_device_interrupt_transfer(usb_device,
+					      self->epin,
+					      data,
+					      CORSAIR_CMD_SIZE,
+					      &actual_len,
+					      timeout,
+					      NULL,
+					      error);
+	if (!ret) {
+		g_prefix_error(error, "failed to get command response: ");
+		return FALSE;
+	}
+	if (actual_len != CORSAIR_CMD_SIZE) {
+		g_set_error(error,
+			    G_IO_ERROR,
+			    G_IO_ERROR_INVALID_DATA,
+			    "wrong size read: %" G_GSIZE_FORMAT,
+			    actual_len);
+		return FALSE;
+	}
+
+	if (data[2] != 0) {
+		g_set_error(error,
+			    G_IO_ERROR,
+			    G_IO_ERROR_FAILED,
+			    "device replied with error: %" G_GSIZE_FORMAT,
+			    data[2]);
+		return FALSE;
+	}
+
+	return TRUE;
+}
+
+static gboolean
+fu_corsair_device_probe(FuDevice *device, GError **error)
+{
+	FuCorsairDevice *self = FU_CORSAIR_DEVICE(device);
+	GUsbDevice *usb_device = fu_usb_device_get_dev(FU_USB_DEVICE(device));
+	GUsbInterface *iface = NULL;
+	GUsbEndpoint *ep = NULL;
+	g_autoptr(GPtrArray) ifaces = NULL;
+	g_autoptr(GPtrArray) endpoints = NULL;
+
+	if (!FU_DEVICE_CLASS(fu_corsair_device_parent_class)->probe(device, error))
+		return FALSE;
+
+	ifaces = g_usb_device_get_interfaces(usb_device, error);
+	if (ifaces == NULL || ifaces->len < 2) {
+		g_set_error_literal(error,
+				    FWUPD_ERROR,
+				    FWUPD_ERROR_NOT_FOUND,
+				    "update interface not found");
+		return FALSE;
+	}
+
+	iface = g_ptr_array_index(ifaces, CORSAIR_VENDOR_INTERFACE_ID);
+	endpoints = g_usb_interface_get_endpoints(iface);
+	/* expecting to have two endpoints for communication */
+	if (endpoints == NULL || endpoints->len != 2) {
```

Comment:
maybe `<= 2` is more future proof for future designs?

Reply:
good idea but we couldn't come up with any idea of how to select correct endpoints from more than 2. But I future proofed some stuff: moved interface number to quirks, changed code to use command sizes from endpoint sizes (in preparation for bigger high speed packets)